source/LifecycleMethod.cpp

```diff
@@ -142,14 +202,20 @@ bool LifecycleMethodCall::operator==(const LifecycleMethodCall& other) const {
 }
 
 LifecycleMethod LifecycleMethod::from_json(const Json::Value& value) {
-  auto base_class_name = JsonValidation::string(value, "base_class_name");
-  auto method_name = JsonValidation::string(value, "method_name");
-  std::vector<LifecycleMethodCall> callees;
-  for (const auto& callee : JsonValidation::nonempty_array(value, "callees")) {
-    callees.emplace_back(LifecycleMethodCall::from_json(callee));
+  std::string base_class_name = JsonValidation::string(value, "base_class_name");
+  std::string method_name = JsonValidation::string(value, "method_name");
+  if (JsonValidation::has_field(value, "callees")) {
+    std::vector<LifecycleMethodCall> callees;
+    for (const auto& callee : JsonValidation::null_or_array(value, "callees")) {
+      callees.push_back(LifecycleMethodCall::from_json(callee));
+    }
+    return LifecycleMethod(base_class_name, method_name, callees);
+  } else if (JsonValidation::has_field(value, "control_flow_graph")) {
+    JsonValidation::validate_object(value, "control_flow_graph");
+    LifeCycleMethodGraph graph = LifeCycleMethodGraph::from_json(JsonValidation::object(value, "control_flow_graph"));
+    return LifecycleMethod(base_class_name, method_name, graph);
   }
-
-  return LifecycleMethod(base_class_name, method_name, callees);
+  throw std::invalid_argument("Invalid JSON format for LifecycleMethod");
```

Comment:
let's be a bit more descriptive, otherwise this will be confusing to the user.
We should also use JsonValidationError.
```suggestion
  throw JsonValidationError(
          value,
          /* field */ std::nullopt,
          "key `callees` or `control_flow_graph`");
```